crates/bevy_pbr/src/render/prepass.rs

```diff
@@ -0,0 +1,570 @@
+
+use crate::{
+    AlphaMode, DrawMesh, Material, MaterialPipeline, MaterialPipelineKey, MeshPipeline,
+    MeshPipelineKey, MeshUniform, RenderMaterials, SetMaterialBindGroup, SetMeshBindGroup,
+    MAX_DIRECTIONAL_LIGHTS,
+};
+
+use std::{hash::Hash, marker::PhantomData};
+
+pub const PREPASS_DEPTH_FORMAT: TextureFormat = TextureFormat::Depth32Float;
+
+pub const PREPASS_SHADER_HANDLE: HandleUntyped =
+    HandleUntyped::weak_from_u64(Shader::TYPE_UUID, 921124473254008983);
+
+pub const PREPASS_BINDINGS_SHADER_HANDLE: HandleUntyped =
+    HandleUntyped::weak_from_u64(Shader::TYPE_UUID, 5533152893177403494);
+
+pub struct PrepassPlugin<M: Material>(PhantomData<M>);
+
+impl<M: Material> Default for PrepassPlugin<M> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+
+impl<M: Material> Plugin for PrepassPlugin<M>
+where
+    M::Data: PartialEq + Eq + Hash + Clone,
+{
+    fn build(&self, app: &mut bevy_app::App) {
+        load_internal_asset!(
+            app,
+            PREPASS_SHADER_HANDLE,
+            "prepass.wgsl",
+            Shader::from_wgsl
+        );
+
+        load_internal_asset!(
+            app,
+            PREPASS_BINDINGS_SHADER_HANDLE,
+            "prepass_bindings.wgsl",
+            Shader::from_wgsl
+        );
+
+        let Ok(render_app) = app.get_sub_app_mut(RenderApp) else {
+            return;
+        };
+
+        render_app
+            .add_system_to_stage(RenderStage::Extract, extract_camera_prepass_settings_phase)
+            .add_system_to_stage(RenderStage::Prepare, prepare_prepass_textures)
+            .add_system_to_stage(RenderStage::Queue, queue_prepass_view_bind_group::<M>)
+            .add_system_to_stage(RenderStage::Queue, queue_prepass_material_meshes::<M>)
+            .add_system_to_stage(RenderStage::PhaseSort, sort_phase_system::<Opaque3dPrepass>)
+            .add_system_to_stage(
+                RenderStage::PhaseSort,
+                sort_phase_system::<AlphaMask3dPrepass>,
+            )
+            .init_resource::<PrepassPipeline<M>>()
+            .init_resource::<DrawFunctions<Opaque3dPrepass>>()
+            .init_resource::<DrawFunctions<AlphaMask3dPrepass>>()
+            .init_resource::<PrepassViewBindGroup>()
+            .init_resource::<SpecializedMeshPipelines<PrepassPipeline<M>>>()
+            .add_render_command::<Opaque3dPrepass, DrawPrepass<M>>()
+            .add_render_command::<AlphaMask3dPrepass, DrawPrepass<M>>();
+    }
+}
+
+#[derive(Resource)]
+pub struct PrepassPipeline<M: Material> {
+    pub view_layout: BindGroupLayout,
+    pub mesh_layout: BindGroupLayout,
+    pub skinned_mesh_layout: BindGroupLayout,
+    pub material_layout: BindGroupLayout,
+    pub material_vertex_shader: Option<Handle<Shader>>,
+    pub material_fragment_shader: Option<Handle<Shader>>,
+    pub material_pipeline: MaterialPipeline<M>,
+    _marker: PhantomData<M>,
+}
+
+impl<M: Material> FromWorld for PrepassPipeline<M> {
```

Comment:
Do you need to impl FromWorld and call init_resource() for the pipeline struct, if you're going to make a specalized pipeline anyways?

Reply:
Nvm, I understand how this works now.